### PR TITLE
TST: Use pytest-localserver instead of making network connections

### PR DESCRIPTION
**Pull request by mroeschke**
Since we have a lot of CI jobs with unit test that make network connections, it is better if the network connections could be made locally if possible. This PR refactors all tests that make network connections to use `pytest-localserver` instead

### CODE REVIEW

File: doc/source/development/contributing_codebase.rst [hunk 1]
```diff
@@ -613,22 +613,15 @@ Testing involving network connectivity
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 It is highly discouraged to add a test that connects to the internet due to flakiness of network connections and
-lack of ownership of the server that is being connected to. If network connectivity is absolutely required, use the
-``tm.network`` decorator.
+lack of ownership of the server that is being connected to. If network connectivity is absolutely required, mock
```
Comment (rhshadrach) [inline doc/source/development/contributing_codebase.rst:616]: With this, are these still highly discouraged? Seems like all the reasons to discourage them no longer apply.

File: pandas/tests/io/excel/test_readers.py [hunk 2]
```diff
@@ -876,19 +876,11 @@ def test_corrupt_bytes_raises(self, engine):
             pd.read_excel(bad_stream)
 
     @pytest.mark.network
```
Comment (rhshadrach) [inline pandas/tests/io/excel/test_readers.py:878]: Similarly, do these need to be marked network anymore? I suppose we're still going over the network, even if it's an internal server, and that could pose issues for different setups?

File: pandas/tests/io/test_s3.py [hunk 3]
```diff
@@ -21,28 +20,31 @@ def test_streaming_s3_objects():
...
+@pytest.mark.single_cpu
+def test_read_without_creds_from_pub_bucket(s3_public_bucket_with_data, s3so):
     # GH 34626
-    # Use Amazon Open Data Registry - https://registry.opendata.aws/gdelt
-    result = read_csv("s3://gdelt-open-data/events/1981.csv", nrows=3)
+    result = read_csv(
+        f"s3://{s3_public_bucket_with_data.name}/tips.csv",
+        nrows=3,
+        storage_options=s3so,
+    )
```
Comment (rhshadrach) [inline pandas/tests/io/test_s3.py:30]: These are still accessing an external s3 bucket, yea? Not sure I understand the change here.

**Review by rhshadrach — COMMENTED** [review]
This looks awesome! Just a few questions.

File: pandas/tests/io/test_s3.py (hunk 3 above)
Comment (mroeschke) [reply pandas/tests/io/test_s3.py:30]: Sorry, this used to be accessing a public bucket, but we have a mock public bucket that we run in our CI (via moto)

File: pandas/tests/io/excel/test_readers.py (hunk 2 above)
Comment (mroeschke) [reply pandas/tests/io/excel/test_readers.py:878]: I decided to keep making these as `network` just for consistency, even as you mentioned its a local network

File: doc/source/development/contributing_codebase.rst (hunk 1 above)
Comment (mroeschke) [reply doc/source/development/contributing_codebase.rst:616]: Good point I'll clarify the wording here

**Review by rhshadrach — APPROVED** [review]
lgtm

**Review by WillAyd — APPROVED** [review]
nice idea

**Comment by phofl** [comment]
This broke the env on ARM Mac. pytest local server requires aiosmtpd which isn't available on Mac

**Comment by WillAyd** [comment]
I am OK with revert unless @mroeschke has an objection

**Comment by mroeschke** [comment]
Could we just remove pytest localserver from environment.yaml? This plugin helps save a few minutes from CI runtime and having the network rests run locally probably isn't a huge priority

**Pull request merged**